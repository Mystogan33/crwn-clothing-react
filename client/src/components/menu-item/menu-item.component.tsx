import React from "react";
import { withRouter, RouteComponentProps } from "react-router-dom";

import {
	MenuItemContainer,
	BackgroundImageContainer,
	ContentContainer,
	ContentTitle,
	ContentSubtitle,
} from "./menu-item.styles";
import { ICategory } from "../../interfaces/interfaces";

type MenuItemProps = Omit<ICategory, "id"> & RouteComponentProps;

export const MenuItemComponent = ({
	title,
	imageUrl,
	size,
	history: { push },
	linkUrl,
	match: { url },
}: MenuItemProps) => (
	<MenuItemContainer
		size={size}
		onClick={() => push(`${url}${linkUrl}`)}
	>
		<BackgroundImageContainer
			className="background-image"
			imageUrl={imageUrl}
		/>
		<ContentContainer className="content">
			<ContentTitle>{title.toUpperCase()}</ContentTitle>
			<ContentSubtitle>SHOP NOW</ContentSubtitle>
		</ContentContainer>
	</MenuItemContainer>
);

export const MenuItem = withRouter(MenuItemComponent);
