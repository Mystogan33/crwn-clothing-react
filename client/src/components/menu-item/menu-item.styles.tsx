import styled from 'styled-components';
import { ICategory } from '../../interfaces/interfaces';

export const MenuItemContainer = styled.div<Partial<ICategory>>`
	height: ${({ size }) => (size ? '380px' : '240px')};
	min-width: 30%;
	overflow: hidden;
	flex: 1 1 auto;
	display: flex;
	align-items: center;
	justify-content: center;
	margin: 0 7.5px 15px;
	border-radius: 0.3rem;
	box-shadow: 0;
	transition: all .4s ease-in-out;
	box-shadow: 0 1px 5px rgba(10, 10, 10, 0.35);

	&:hover {
		cursor: pointer;
		box-shadow: 0 1px 10px rgba(10, 10, 10, 0.7);

		& .background-image {
			transform: scale(1.1);
		}

		& .content {
			opacity: 0.9;
		}
	}

	&:first-child {
    margin-right: 7.5px;
  }
  &:last-child {
    margin-left: 7.5px;
  }

	@media screen and (max-width: 800px) {
		height: 200px;
	}
`;

export const BackgroundImageContainer = styled.div<Partial<ICategory>>`
  width: 100%;
  height: 100%;
  background-size: cover;
  background-position: center;
  background-image: ${({ imageUrl }) => `url(${imageUrl})`};
	transition: transform 3s cubic-bezier(0.25, 0.45, 0.45, 0.95);
`;

export const ContentContainer = styled.div`
  height: 90px;
  padding: 0 25px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  border: 1px solid black;
  background-color: white;
  opacity: 0.7;
  position: absolute;
  border-radius: 0.2rem;
	transition: opacity .3s ease-in-out;
`;

export const ContentTitle = styled.span`
  font-weight: bold;
  margin-bottom: 6px;
  font-size: 22px;
  color: #4a4a4a;
`;

export const ContentSubtitle = styled.span`
  font-weight: lighter;
  font-size: 16px;
`;
