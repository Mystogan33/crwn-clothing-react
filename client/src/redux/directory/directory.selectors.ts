import { createSelector } from 'reselect';
import { AppState } from '../root-types';

const selectDirectory = (state: AppState) => state.directory;

export const selectDirectorySections = createSelector(
  [selectDirectory],
  directory => directory.sections
); 
