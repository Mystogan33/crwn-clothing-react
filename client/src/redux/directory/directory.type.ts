import { ICategories } from "../../interfaces/interfaces";

export interface DirectoryState {
  sections: ICategories
};