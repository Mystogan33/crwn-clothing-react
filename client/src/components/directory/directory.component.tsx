import React from "react";

import { selectDirectorySections } from "../../redux/directory/directory.selectors";
import { useAppSelector } from "../../redux/hooks";

import { MenuItem } from "../menu-item/menu-item.component";
import { DirectoryMenuContainer } from "./directory.styles";

export const Directory = () => {
	const sections = useAppSelector(selectDirectorySections);
	const renderedMenuItems = sections.map(({ id, ...othersProps }) => (<MenuItem key={id} {...othersProps} />));

	return (
		<DirectoryMenuContainer>
			{renderedMenuItems}
		</DirectoryMenuContainer>
	)
};
