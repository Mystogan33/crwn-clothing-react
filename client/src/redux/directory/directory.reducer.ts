import { DIRECTORY_DATA } from './directory.data';
import { DirectoryState } from './directory.type';
import { Action } from 'redux';

const INITIAL_STATE: DirectoryState = {
  sections: DIRECTORY_DATA
};

const directoryReducer = (state = INITIAL_STATE, action: Action) => {
  switch(action.type) {
    default:
      return state;
  }
};

export default directoryReducer;
