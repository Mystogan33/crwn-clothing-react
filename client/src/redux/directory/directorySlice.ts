import { createSlice } from '@reduxjs/toolkit';
import { ICategories } from '../../interfaces/interfaces';
import { DIRECTORY_DATA } from './directory.data';

interface DirectoryState {
  sections: ICategories
};

const initialState: DirectoryState = {
  sections: DIRECTORY_DATA
};

export const directorySlice = createSlice({
  name: 'directory',
  initialState,
  reducers: {}
});

export default directorySlice.reducer;