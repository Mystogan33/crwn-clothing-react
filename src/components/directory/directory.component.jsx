import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import { MenuItem } from '../menu-item/menu-item.component';

import { selectDirectorySections } from '../../redux/directory/directory.selectors';

import { DirectoryMenuContainer } from './directory.styles';

const mapStateToProps = createStructuredSelector({
  sections: selectDirectorySections
});

export const Directory = connect(mapStateToProps)(
  ({sections}) => (
    <DirectoryMenuContainer>
      { sections.map(({id, ...othersProps}) => (
          <MenuItem key={id} {...othersProps} />
        ))
      }
    </DirectoryMenuContainer>
  )
)
