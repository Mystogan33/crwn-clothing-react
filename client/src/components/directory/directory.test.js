import React from 'react';
import { shallow } from 'enzyme';
import { DirectoryComponent } from './directory.component';
import { DIRECTORY_DATA } from '../../redux/directory/directory.data';
import { MenuItem } from '../menu-item/menu-item.component';

describe('Directory component', () => {
  let wrapper;
  let mockSections;

  beforeEach(() => {
    mockSections = DIRECTORY_DATA;
    wrapper = shallow(<DirectoryComponent sections={mockSections} />);
  });

  it('should render Directory component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a number of MenuItem equal to sections prop length', () => {
    expect(wrapper.exists(MenuItem)).toBe(true);
    expect(wrapper.find(MenuItem).length).toEqual(mockSections.length);
  });

  it('should not render any MenuItem if sections is empty', () => {
    const newWrapper = shallow(<DirectoryComponent sections={[]} />);
    expect(newWrapper.exists(MenuItem)).toBe(false);
  });
});