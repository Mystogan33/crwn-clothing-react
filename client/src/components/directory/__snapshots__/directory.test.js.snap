// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Directory component should render Directory component 1`] = `ShallowWrapper {}`;
